File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.js
function isPrime(number) {
    if (number < 2) return false;
    for (let i = 2; i * i <= number; i++) {
        if (number % i === 0) return false;
    }
    return true;
}

function getPermutation(arr, number) {
    let result = [];
    const visited = Array(arr.length).fill(false); 
    
    function dfs(current) {
        if (current.length === number) {
            result.push([...current]);
            return;
        }
        for (let i = 0; i < arr.length; i++) {
            if (!visited[i]) {
                visited[i] = true;
                current.push(arr[i]);
                dfs(current);
                current.pop();
                visited[i] = false;
            }
        }
    }

    dfs([]);
    return result;
}

function solution(numbers) {
    let answer = 0;
    const numberSet = new Set();
    const numArr = [...numbers];

    for (let i = 1; i <= numArr.length; i++) {
        const permutations = getPermutation(numArr, i);
        permutations.forEach(numArr => {
            const num = Number(numArr.join(''));
            numberSet.add(num);
        });
    }
    numberSet.forEach(num => {
        if (isPrime(num)) answer++;
    });

    return answer;
}

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
from math import isqrt

def is_prime(number):
    if number < 2:
        return False
    for i in range(2, isqrt(number) + 1):
        if not number % i:
            return False
    return True

def solution(numbers):
    n = len(numbers)
    visited = [False] * n
    # 소수의 중복을 허락하면 안되니까 집합 사용
    prime_set = set()
    numbers = list(numbers)
    
    def backtrack(number_str):
        # 빈문자열이 아닌 경우
        if number_str:
            # 숫자로 형변환
            number = int(number_str)
            # 소수 검사
            if is_prime(number):
                prime_set.add(number)
        # 한번 조합으로 가능한 모든 수 도달 시 돌아가기
        if len(number_str) == n:
            return
        
        for i in range(n):
            if not visited[i]:
                visited[i] = True
                backtrack(number_str + numbers[i])
                visited[i] = False
    backtrack('')
    
    return len(prime_set)
                